Rechunk copy feature

dgkmd opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->

It supports `copy --rechunk` feature which rechunks data while copying snapshots between repositories, as discussed in #5473.

The code consists of main `rechunker` module and a few submodules as follows: blob cache, scheduler, and worker; along with modification to `cmd_copy.go`.

Main `rechunker` module handles overall workflow, that is, to rechunk-copy snapshots. 

Blob cache acts as a transparent cache layer of the source repo. It is needed because the blobs are scattered across packs non-contiguously while rechunk process needs stream-reading the file.

Scheduler manages file processing order. Specifically, it prioritizes files whose all consisting blobs are ready in the blob cache, thus minimizing redundant pack download.

Worker rechunks each assigned file. It can be run multi-threaded and also can run standalone without upper layer modules, if needed (e.g., to rechunk single file).

The feature can be used with `copy --rechunk` command. It has two additional arguments on top of plain `copy`. `--cache-size` controls maximum blob cache size in MiB, and `--add-tag` adds tags of the names specified to the created snapshots.

For safety, `copy --rechunk` can be run only if `RESTIC_FEATURES=rechunk-copy` environment variable is set.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Closes #5473 
https://forum.restic.net/t/is-it-possible-to-re-chunk-after-a-restic-copy/6072
https://forum.restic.net/t/copy-snapshots-between-repositories-without-copy-chunker-params/7320


Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

wplapper commented:
Hi dgkmd,
You are still suffering from the same performance problem which I discovered earlier during my tests. In today's test I use a local (empty) repository as target and rclone:onedrive:xxx as my source repository. I try to copy a small snapshot  with ~4,000 data blobs and a size of uncompressed 41.6 MiB, compressed 10.7 MiB. My read performance is about 7 MiB/s, so the reading of all blobs should take a few seconds.  The normal copy takes all in all ~13 seconds, and I rechunk-copied about 60% after 1,000 seconds. I will have a closer look later where the command gets stuck.

dgkmd commented:
Hi @wplapper ,
Thank you for your thorough feedback.

Pack cache works as follows.
1. When a blob `A` is requested and the blob is *not* in the cache:
    1) Locates the pack `P` in which the blob `A` resides.
    2) Downloads all blobs in the pack `P` that are referenced by at least one file, with `Repository.LoadBlobsFromPack`.
    3) Stores the blobs to the cache.
    4) If the number of stored **packs** exceeds LRU_SIZE, it evicts the oldest pack `P'`. Which means, it evicts all blobs loaded from `P'` from the cache.
    5) Return `A` through the async channel.
2. When a blob `A` is requested and the blob *is* in the cache:
    1) Updates LRU recency of the pack `P` in which the blob resides.
    2) Immediately returns the blob `A`.

Additionally, it implements scheduling algorithm where it prioritizes processing the files whose blobs are all ready in the cache.

The memory usage is proportional to the pack cache size, which is hard coded in LRU_SIZE variable. The empirical memory usage estimation is `25 MiB * LRU_SIZE`, which is about 5 GiB in current setting (`LRU_SIZE = 200`). So at least the cache size will not explode with the repo size :)
Determining LRU_SIZE is indeed a trade-off problem. There will be no eviction if LRU_SIZE is very large, so no re-downloading packs, at the cost of huge memory consumption (and vice versa).

On handling single-blob files, we cannot guarantee it will be single-blob when chunked with different chunking parameters. Also we cannot know whether it will be chunked single-blob ahead of time. Therefore I think it's our best to just process all files...

I'll add plenty of `debug.Log()` for traceability, following your advice.
And also make no-more-needed-blobs evicted immediately and compare the performance.

Thank you again for your considerate review!

Best regards,

dgkmd commented:
Hi wplapper, 
Thanks for the comment.

I updated the code, where I split `RechunkData()` to smaller sub-functions. Now I'm working on improving the cache considering your comments. It may take a while.
Any further feedbacks are appreciated!

wplapper commented:
Hi [dgkmd](https://github.com/dgkmd),
I copied your code and added more debugging output to your code. I wanted to understand what is happening with all these many parallel pipelines. What I have seen before and and saw again that all these pipelines get stuck once in a while and nothing happens during this time. I would like to attach a compressed restic-debug.log.xz, but github won't let me do that. Could you please send me your email address and I will forward these files to you. 

But I will give you one of many more examples in the log:
```
2025/10/18 18:35:41.855006 index/index.go:118	index.init.func1	91	index 0xc01cdd2388 only has 3989 blobs and is too young (29.225628461s)
2025/10/18 18:35:41.855017 index/master_index.go:219	index.(*MasterIndex).finalizeFullIndexes	91	index 0xc01cdd2388 not full
2025/10/18 18:35:41.855026 index/master_index.go:223	index.(*MasterIndex).finalizeFullIndexes	91	return 0 indexes
2025/10/18 18:35:41.958217 debug/round_tripper.go:110	debug.loggingRoundTripper.RoundTrip	199	------------  HTTP RESPONSE ----------
HTTP/2.0 206 Partial Content
Content-Length: 17359408
Accept-Ranges: bytes
Content-Range: bytes 0-17359407/17360838
Content-Type: application/octet-stream
Date: Sat, 18 Oct 2025 17:35:41 GMT
Last-Modified: Mon, 25 Aug 2025 02:35:02 GMT
Server: rclone/v1.71.1

2025/10/18 18:35:43.632228 logger/log.go:47	logger.(*Backend).Load	198	  load err <nil>
2025/10/18 18:35:43.632280 repository/repository.go:1165	repository.(*packBlobIterator).Next	198	  process blob <data/73625e26>, skipped 0, <Blob (data) 73625e26, offset 0, length 341562, uncompressed length 647512>
2025/10/18 18:35:43.637437 repository/repository.go:1165	repository.(*packBlobIterator).Next	198	  process blob <data/93e955c7>, skipped 0, <Blob (data) 93e955c7, offset 341562, length 86482, uncompressed length 93442>
2025/10/18 18:35:43.637735 repository/repository.go:1165	repository.(*packBlobIterator).Next	198	  process blob <data/a3ff028d>, skipped 0, <Blob (data) a3ff028d, offset 428044, length 2206576, uncompressed length 2314775>
2025/10/18 18:35:43.643707 repository/repository.go:1165	repository.(*packBlobIterator).Next	198	  process blob <data/a4992818>, skipped 0, <Blob (data) a4992818, offset 2634620, length 86425, uncompressed length 93428>
```




There is a time difference of about 1.5 seconds before the system wakes up again. I feel, this is not right. You seem to put all your eggs in one basket, and you have to wait until this HTTP request is satisfied. Everything else came to a standstill.

I will attach the modified source code
[rechunker_wpl.txt](https://github.com/user-attachments/files/22986683/rechunker_wpl.txt)
rename file to .go again ...

dgkmd commented:
Hi @wplapper ,
I reimplemented the blob cache and also updated other code. Major changes are as follows:
1) Blobs are loaded by the pack, but evicted by each blob.
2) Rechunker traces blob usage and notifies the cache to ignore no longer needed blob (makes it evict the blob immediately and don't download it anymore)
3) Files for rechunking are ordered by their length (chunk count) -longest first- as you had commented in the email.
4) The cache is now enabled by default, and you can set its size by giving an argument `--cache-size` in MiB unit.

From my observation, redundant blob download to some extent were inevitable, especially if packs were shaped by interleaving multiple files. Redundant downloads were minimized when I set a huge cache size (>10 GiB), but currently I set default cache size to 4096 MiB to make it safe for majority of users.

Please let me know if the code works as you expect.

Best regards,

wplapper commented:
Hi @dgkmd,

Thank you very much for your most recent update to your rechunker code. On the whole it looks very good, but you still have a problem with dispatching the ``rc.priorityFilesList`` quickly enough. Dispatching these readily available files will remove a lot of memory pressure from the execution of your command. In the example snapshot I used, here the summary
```
   3823 unique files in selected snapshots
      6 regular file downloads to start
   3817 fastTrack files to process
     30 data packfiles to load
   4222 used blobs with size 471.714 MiB

```

there is a build-up of over 2,500 ready files before ``runDispatcher()``can get hold of them because all the other files stuck in the ``chDispatch`` queue. In my opinion, you should have two Go routines in ``runDispatcher()``, one for the normal stuff and one which exclusively deals with new arrivals in ``rc.priorityFilesList``.

The other comment I would like to make: leave the debug.Log() messages in your code, so one can actually see what is happening and when it is happening. Only by using debug.Log() records I could detect your scheduling issue.

I will upload my modification to your code onto my restic github repository. I will leave another comment where to find the code once it has been uploaded.

The attached file abbr_``restic-debug.txt` contains a restic debug log filtered by ``rechunker/rechunk`` records. You can see when new files arrive for dispatch, when the packfiles start to download and when they become ready. Once a file is finished completely, you will see the message ``ended fileID caffeeee``. The priority queue message is ``the prio queue now contains 2565 files``.

[abbr_restic-debug.txt](https://github.com/user-attachments/files/23325290/abbr_restic-debug.txt)

wplapper commented:
Hi @dgkmd,

you can find my modifications to your code here. I also have added a toy to simulate part of the workings of your real code. 

https://github.com/wplapper/restic/tree/rechunker

dgkmd commented:
Hi @wplapper,

Thank you for your review. I looked into your code and debug log. It seems the reason for thousands of `fastTrack` files stacking is that all rechunk workers are still working on non-prioritized large files at that point. So I reimplemented dispatcher logic to serve two channels, as you have commented, one for regular files and one for prioritized files (fast track). Then I added a dedicated rechunker worker that only listens to prioritized files channel. It will start on processing prioritized files as soon as a file is sent to `chPriority`, even when all other workers are busy processing regular files.

Until now I focused on implementing the code, so I didn't take enough care for my code to be readable and easily analyzable. Provided the core logic will not change much from now on, I will add sufficient comments and logs so that anyone can easily analyze the code. Thanks for the advice.

Please let me know if it works as you expected.

Best regards,

wplapper commented:
Hi @dgkmd,

Thanks for the latest update. The first time I ran your new version, I got panic dump. I cannot reproduce it on will, but I have seen a similar one before. Here is the the screen capture from the terminal: 
[panic-001.txt](https://github.com/user-attachments/files/23610111/panic-001.txt)

Otherwise the code looks good, meaningful debug.Log() output!

dgkmd committed:
Squashed commit of the following:

commit 0e6e1d21526387212148c7cfdec8ddf2f3c3150a
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Dec 4 18:25:03 2025 +0900

    WIP

commit e7a1b5c14d0a6214a6f9e76a396c5dcb932e6488
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Dec 4 18:01:47 2025 +0900

    WIP

commit 4c803f9fabe112571642dd99cd89681662858918
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Dec 4 17:24:49 2025 +0900

    WIP

commit 2779f005a955a26a8770caf32c6481ae23223126
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Dec 2 22:17:20 2025 +0900

    WIP: Remove ChunkDict

commit 2f0cdf16d149257d83be3d3ddfcb900edb684b59
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Dec 2 18:54:52 2025 +0900

    WIP

commit fb0580c80b88a20261be1411305d29823f283734
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Dec 2 18:34:11 2025 +0900

    WIP

commit 51168307869c3d62a7f27f13e3fd766ee962e12b
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Dec 2 18:03:49 2025 +0900

    WIP

commit 4a8ccb3e6e32d8feec38373f2406b2b2f84c0dda
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Dec 2 17:30:17 2025 +0900

    WIP

commit 36efa22c492bf4c4fcad3e47a6b1751c3c040d00
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Dec 2 13:38:12 2025 +0900

    WIP

commit 33e85c8ffea90318322b438e2fc81b449901ba9b
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Nov 28 19:01:44 2025 +0900

    WIP

commit 5c23c1d0628b1b612159c94446c1a0f2c0ce5902
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 26 17:55:12 2025 +0900

    WIP

commit 661df4b48b35b252e0e41aaf8296b1aa70994923
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 19 16:48:31 2025 +0900

    Fix: debugNote concurrency issue

    Fix debugNote concurrent r/w issue by implementing dump function

commit 8f28da4c742be41d5bb6af39578b5e760f7909cb
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Nov 18 21:12:59 2025 +0900

    Refactor

    Change code order for better readability
    Change debugNote to a separate struct

commit e9ba3ca89b26ee6eb7c69bc898d757136ee0af1b
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Nov 18 19:41:01 2025 +0900

    Refactor: rechunk_worker

    Integrate largely redundant functions into one
    Introduce chunkDictHelper for encapsulation
    Improve data structures to reduce redundancy and complexity

commit 034e1cb410995b4ea260fed2e66eabe45a42cd84
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Nov 18 01:51:36 2025 +0900

    Docs: Update docs on rechunk-copy

commit f159887d338e3c24ecdc1c1dc54b858d6bec37be
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Nov 18 00:31:33 2025 +0900

    Update: use WithBlobUploader

    Use WithBlobUploader introduced in latest restic update
    Make Progress to preserve final output
    Misc code cleanup (ctx names)

commit f12d17943b71de60a7ef5fbbc73fc3c8e1f23c75
Merge: bdf4737ac 382616747
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Nov 17 20:13:05 2025 +0900

    Merge remote-tracking branch 'origin/master' into feature/rechunker

commit bdf4737acd93ebf39d3243ea46dba363102f3d71
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Nov 17 20:10:13 2025 +0900

    Refactor: rechunk_cache

    Refactor rechunk_cache.go
    Add comments for rechunker package code

commit 66e0a0c0a86c2a2c5757b6060810740d20870e4e
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Nov 17 17:47:50 2025 +0900

    Debug: Add debug logs

    Add debug logs in rechunker_worker.go

commit 3d6ed46b673bd6be797be4b33f8de5410f05b3f1
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Nov 17 17:14:21 2025 +0900

    Feat: Progress bar

    Enrich information in the progress bar

commit 1db294dd3f9e07b0b9526e836bed772a6182689f
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 12 18:08:26 2025 +0900

    Refactor: ChunkDict

    Refactor ChunkDict code

commit 41d4a9c6497ef6d906b3789ab63b1f5ae2d852d2
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 12 17:36:38 2025 +0900

    Fix: Add condition test

    Add condition test for priorityFilesHandler.Push() on blob ready

commit 60b7be1e799db81c2ee173be102b287d8cf8e5e3
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 12 16:36:34 2025 +0900

    Style: golangci-lint

    Comply with golangci-lint suggestion

commit be0a5934298fcd68d8c3cf6f47016450cfee5b2d
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 12 15:56:30 2025 +0900

    Misc: Add RechunkRepo interfaces

    Add RechunkSrcRepo and RechunkDstRepo interfaces to make it compatible with rechunker_test

commit dcab416d682c7dc5aeb34b1debf0d05826abe99a
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Nov 12 15:23:22 2025 +0900

    Refactor

    Reorder functions from broader functions to finer functions
    Split into subfunctions for long function code

commit 1934b88ab3ddeeca3dcc4fadca8ed5b67b558f62
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Nov 11 17:54:49 2025 +0900

    Feat: Ignore no more needed blob after blob load

    Make blob cache ignore no longer needed blobs immediately after blob load
    Refactor code

commit fdc08c86d197088cfd8fb695c8c192c8835c320c
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Nov 10 18:14:04 2025 +0900

    Refactor: Enhance code readability and modularity

    Modularize subfunctions so that they are more readable
    Place 'big' functions first, to make it easy to get the over logic
    Move types and structs to appropriate file
    Change numWorkers to `min(GOMAXPROCS(0), srcRepo.Connections())`

commit 0ff40fb9b39da5ac122738956be598a10f48514f
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sun Nov 9 16:24:23 2025 +0900

    Misc: Change numDownloaders

    Change numDownloaders to `min(numWorkers, srcRepo.Connections())`
    Remove unnecessary PackedBlobLoader interface
    Change HashOfIDs to public function

commit 4cd013e4fa672bdcf5dbce59e2d7e2cd8aeb2bc4
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Nov 7 15:17:30 2025 +0900

    Feat: Split dispatcher channel

    Split chDispatch into chRegular and chPriority
    Introduce PriorityFilesHandler, replacing inline concurrency logic of priorityFilesList
    Refactor names and file orders
    Move rechunk worker code to a separate file

commit fa00c151acb7a690762367112799e5620f085294
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 31 17:53:24 2025 +0900

    Misc: Change numWorkers

commit ddae794c4399f40102d542ac4724cde8e6c0e8f8
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 31 17:34:53 2025 +0900

    Update: Align with global interface

commit 809d1c4aedebdd3080a4dda551846c02b2a02da3
Merge: 08b707b16 71432c7f4
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 31 17:20:45 2025 +0900

    Merge remote-tracking branch 'origin/master' into feature/rechunker

commit 08b707b1628965872a97f58e25627a7d742bfce4
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Oct 29 16:31:54 2025 +0900

    Feat: Reimplement rechunk blob cache

    Rechunk blob cache change including:
      - Blobs loaded by the pack, evicted by the blob
      - Ignore no longer needed blobs
    Order files by their blob count, descending order
    Rechunk blob cache is enabled by default, and its size is customizable
    Refactor code

commit 224b5ca711d0143c61caab36c499179ba9ec0c2c
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Oct 28 00:04:43 2025 +0900

    Feat: Rechunk file order

    Change to process files by lexicographic pack order
    Refactor names

commit 4c18ebb2f9a225832cd56780689be39a702116f0
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Oct 16 20:41:33 2025 +0900

    Refactor: Improve modularity of RechunkData

    Split RechunkData to smaller units
    Rename types and variables
    Remove unnecessary custom types

commit fe5ef9ba57229de33cd661c7629a70d8353d29bd
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 10 20:43:31 2025 +0900

    Debug: cmd_rechunk_copy log

    Add debug.Log for cmd_rechunk_copy.go
    Ensure concurrency safety for rechunker.debugNote
    Fix debug.Log in rechunk_cache.go to log only real requests

commit 4064f7cbc03e1e9f7e98b2e2ce9a7917fa5f32c4
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 10 18:52:34 2025 +0900

    Update: new restic version

    Update interfaces to match the new restic version's

commit cb1982a6eb6538199d87c998f41bd83a931dc8ad
Merge: ef6d24aaf 481fcb9ca
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 10 18:16:31 2025 +0900

    Merge branch 'master' into feature/rechunker

commit ef6d24aaf697b7f43f05ddbf602685ec785e0296
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Oct 10 16:49:14 2025 +0900

    Debug: Add debug trace

    Add debug trace snippets for PackCache and ChunkDict
    Change back numWorkers to GOMAXPROCS(0)

commit 139eaf217acb660f86f13be12b821659b01d0ca6
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Sep 25 14:57:03 2025 +0900

    Misc: Tune hyperparameters

    Tune numWorkers and numDownloaders
    Fix typo in rechunker_test

commit 0d7b12785bcefc0a41a73d698f9d18ab7f38ccc3
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 24 22:10:46 2025 +0900

    Update: Terminal progress printer

    Update cmd_rechunk_copy code to use latest terminal progress printer (as in cmd_copy)
    Remove unnecessary code in cmd_rechunk_copy

commit 30dc52539b9b0e166f7dfb0c0bc6051f99e78d11
Merge: 9a152ae70 3e6b5c34c
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 24 21:42:47 2025 +0900

    Merge remote-tracking branch 'origin/master' into feature/rechunker

commit 9a152ae70d641d9b667b9ffd82bf83c0e4bcb852
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 24 21:32:32 2025 +0900

    Fix: buildIndex

    Fix issue that the needed metadata is not computed when not using PackCache.
    Refactor names and struct fields
    Change test cases to ensure testing ChunkDict

commit 78b9ae5f51c14e9ad736657fdbe4bf7ce2d400f8
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 24 17:54:26 2025 +0900

    Misc: Update changelog

commit da37aa834ca9df21ce8323bdfec05534db122e45
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Sep 18 18:30:11 2025 +0900

    Feat: Pack cache switch

    Add switch for enabling pack cache
    Refactor names
      RechunkBlobCache -> PackCache
      RechunkChainDict -> ChunkDict
    Tune hyperparameters

commit f7fe0952fdd8fa0c31faa8c0eb6bfa14461c7959
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 17 13:47:52 2025 +0900

    Misc: changelog

commit 3bfe8b8b8e688a11841228a7a8ff15a429675ff0
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Sep 15 17:39:51 2025 +0900

    Test: rechunk-copy integration test

    Add cmd_rechunk_copy_integration_test
    Refine rechunker test cases
    Refactor minor names in rechunker and rechunker_test

commit e0635df2599e555d078f322fc43ade2f8ce124b5
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 10 22:30:22 2025 +0900

    Test: rechunker_test

    Augment test cases for RechunkData
    Fix concurrency error in test code
    Fix cmd_rechunk_copy import package names
    Change hyperparameters
    Minor refactoring

commit d6034d4ef95c56b416a1b83adc862c9e7338575f
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 10 21:42:06 2025 +0900

    Rename: Move rechunker to internal/rechunker

    Move rechunker module from internal/walker to internal/rechunker
    Change package name to rechunker

commit 067c423cf46e452df9afbfc7e6c711a324da0b5e
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Sep 10 19:35:04 2025 +0900

    Feat: rechunker_test

    Rewrite rechunker_test for current rechunker
    Refactor method signatures for modularity

commit 75522f91879df44a14f366263815e1b6efe65e12
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Sep 8 20:36:16 2025 +0900

    Style: golangci-lint

    Fix code style to comply with golangci-lint
    Add comments and refactor names
    Tune hyperparameters

commit 4180a847ccf4232994f265653faba33bd592c224
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Sep 8 16:09:43 2025 +0900

    Feat: Remove blob tracer

    Remove blob tracer, as no performance difference in ablation study

commit 10fc276a63cc5ed3c4798615ae60b3be709a92c9
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Sep 8 15:43:33 2025 +0900

    Feat: numDownloaders

    Implement multiple downloaders in blob cache

commit e73749cbb0644efd8622123c9dfa1f7d3d4a2a4a
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sun Sep 7 23:38:35 2025 +0900

    Feat: Blob tracer

    Reintroduce blob tracer
    Refactor names and code structure

commit 3451587b521428f26b6cd3c9ab4ecf15d7ca04cb
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Sep 2 00:25:17 2025 +0900

    Feat: RechunkChainDict

    Reimplement Rechunker to accommodate RechunkChainDict
    Refactor names, types, and structures

commit b7c9be234e68d40c889c7041cf1612418ae35ac7
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sun Aug 31 02:22:24 2025 +0900

    Misc: Rechunker jump feature plan

commit 2c1328fb10e2161de9171e5fd802802c2b48ffb3
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sun Aug 31 00:21:22 2025 +0900

    Feat: RechunkChainDict

    Implement RechunkChainDict.Add
    Refactor names

commit a1101c0fb6e73ce5852f8cc48333735eb73d93a6
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sat Aug 30 19:41:09 2025 +0900

    Feat: RechunkChainDict

    Implement RechunkChainDict
    Refactor names

commit 393603d9cf11fb27f251d67321724b814d0e7449
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Aug 28 22:30:12 2025 +0900

    Feat: Optimize scheduling

    Add optimization, such as:
      Prioritize files whose blobs are all ready in BlobCache
      Don't save to BlobCache no longer needed blobs
    Refactor names and

commit 3b04337ebbdd7b7a2e99253be1e3ea7668e986ba
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Aug 27 20:39:46 2025 +0900

    Feat: Rechunk copy with caching and scheduling

    Implement PackLRU-based BlobCache for Rechunker
    Reimplement and refactor Rechunker

commit 2278a83c6007ef2888c5d4d92096606201e1fa3b
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Aug 26 20:49:14 2025 +0900

    misc

    Reimplementing rechunker

commit bf071940d05f0fe6ac298619d8af5c948190f12c
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Aug 25 16:58:16 2025 +0900

    misc

commit 1e14b3eaf8d035079ee0e7a354fac0c0927aee6f
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Aug 25 09:22:30 2025 +0900

    Fix: bufferPool implementation (#5473)

    Increase bufferPool capacity (2 -> 4) per worker
    Move bufferPool to global (shared among workers)

commit eab0e1d3ca3fb57249f7af3df902951768ad317c
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sat Aug 23 00:43:16 2025 +0900

    Style: gofmt

commit ff70fb86081a9c953feef04798d54ef4af519a0b
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Aug 22 23:12:28 2025 +0900

    Fix: Potential memory leak

    Fix issues addressed in review by wplapper (#5473)
      Fix buffer not returning to bufferPool at the end of file
      Change numWorkers value to runtime.GOMAXPROCS(0) to utilize multicore

commit 92e37c1f8ae13cfedc4bfced82ef0dc988c5054b
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Aug 19 18:22:26 2025 +0900

    Style: Cleanup code

    Cleanup and organize rechunk code
    Add feature flag check in rechunk-copy command

commit 619968326a99e2ea588705d99f4a61b5d9ec3bca
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Aug 19 18:22:20 2025 +0900

    Feat: Add feature flag for rechunk-copy

commit c82f793b90b1365aeb1a9acec2980fd780480689
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Aug 14 18:07:33 2025 +0900

    Misc: Minor refactoring

commit eaf6534487b354491132d9b217bba798757d57cb
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Aug 4 17:47:37 2025 +0900

    Misc: Change config value

    Change default numWorkers of RechunkData to 1
    Misc UI message change

commit 25e562b3a663894c8ac18f69612d1b75b1b540f1
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Aug 4 13:37:35 2025 +0900

    Fix: Map concurrency error of rechunkBlobsMap

    Fix map concurrency error by adding mutex to rechunkBlobsMap
    Remove bloblru from rechunker
    Add buffer to downloader worker of RechunkData
    Simplify rechunkBlobsMap's value
    Add tags feature to rechunk-copy

commit da48871d1aaa6a45b230a47af5f636388d791253
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Jul 30 11:40:10 2025 +0900

    Feat: Add progress bar

    Add progress bar to rechunk-copy
    Minor refactoring

commit 0d7da20db671e26e020a890cb4131ff1d3517cad
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Wed Jul 30 00:37:21 2025 +0900

    Feat: Integrate to cmd

    Implement cmd_rechunk_copy.go
    Check working on the built program with example repos

commit 0e472bcb5916f3201b3747990231e75cd1ca85ae
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Tue Jul 29 22:38:11 2025 +0900

    Feat: TestRechunkRewriter

    Implement TestRechunkRewriter
    Refactor rechunker.go

commit 75c5112caeda9c31105a2d5f7b0c31badd7af8bb
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Jul 28 23:55:03 2025 +0900

    Fix: Debug and code cleanup

commit 0a61d3c7bcbf1e6ca68c791f6ad203dd7eb8cf45
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Mon Jul 28 13:08:22 2025 +0900

    Feat: Implement TestRechunker

commit b38f4c06b8160e6b08509581620ba6160e277176
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Sun Jul 27 20:48:00 2025 +0900

    Misc: Write plan of rechunker_test.go

commit 0ce37d8d5c79de8455cf603f2db1024a4378c7c0
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Jul 25 23:39:21 2025 +0900

    Feat: Rechunker worker

    Implement rechunk worker

commit 9d1afe0373ba8a85f955017be393a3bec95f722d
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Fri Jul 25 17:23:29 2025 +0900

    Feat: rechunker tree traversal

    Implement traverser in FileRechunker that registers unique file contents to the rechunk job chan
    Implement tree rewriter that substitutes Content field to rechunked blobs

commit 8419d0991b5c552ac81d80b7584a28cd3a440c9f
Author: Donggyu Kim <kimdonggyu.dev@gmail.com>
Date:   Thu Jul 24 19:29:00 2025 +0900

    misc

dgkmd committed:
Update: Align function signature

dgkmd commented:
Hi @wplapper,

It's been a while since last update.
While refactoring the overall code, I felt that the existence of ChunkDict feature makes the entire code significantly more complex, although it can boost performance in specific scenarios.
I thought that it would harm long-term maintainability of the code, so I got rid of ChunkDict from the code.
Also, the worker pipeline is simplified; it now consists of only reader and writer.

It is quite an overhaul of the code, but I focused on simplifying and disentangling the code. Please have a look when you have time.

Best regards,

wplapper commented:
Hi dgkmd,
I will have a look next week. We are currently visiting our youngest son and his family.

dgkmd commented:
It's nice. Have a good time!

dgkmd committed:
Refactor: Minor refactoring and code cleanup

dgkmd committed:
Change rechunker_test code

dgkmd committed:
Feat: Track all files' prefix

Track all files' prefix to discover priority files,
instead of tracking only small files.

dgkmd commented:
Code update status: Tracking blob load status for prioritized process is now done for prefixes of all files, instead of being done only for small files. 
There is about 10% performance improvement, when tested with a local, small sized repository(<100 GiB).

dgkmd committed:
Misc: Remove redundant debug log

dgkmd committed:
Fix: Blob cache concurrency bug

Fix potential race condition inside blob cache downloader code

dgkmd committed:
Style: gofmt and golangci-lint

Fix the code to follow golangci-lint suggestion

dgkmd commented on internal/rechunker/dispatcher.go:
```diff
@@ -0,0 +1,236 @@
+	}
+
+	select {
+	case <-d.done:
```
It is a trick to avoid double-closing d.done.

dgkmd committed:
Refactor: Remove unnecessary variables

dgkmd committed:
Fix: Dispatcher last file bug

Fix dispatcher bug in which the last file of the priority list may be omitted.

dgkmd committed:
Misc: Add potential error handling logic

Add error handling logic for situations where a wrong blob id is given to BlobCache.Get().

dgkmd committed:
Fix: BufferPool capacity

Increase BufferPool capacity to 3x of numWorkers.

MichaelEischer commented:
Thanks for working on this feature! I just had a very, very brief look at this PR and I've already noticed a few larger problems:
- Why is this a separate `rechunk-copy` command and not integrated with `copy --rechunk`? This just clutters the CLI.
- The PR is far larger than I'd like. It also lacks any integration with existing code and adds yet another pipeline to download and upload files. As part of https://github.com/restic/restic/issues/2446, I'm currently working on adding `Object{Reader,Writer}` types which would allow rechunking an individual file using just `io.Copy(objectWriter, objectReader)`. I've seen that `worker.go` contains some related code, which will have to be replaced once I'm done. My hope is also that we can unify the code of `ObjectReader`, `internal/dumper` and the `BlobCache` from this PR. Please don't make any major changes in that regard until my PR is far enough.
- I hope we can get rid of the dispatcher, although I haven't looked into the code far enough to tell whether that's actually an option.
- `Rechunker.RewriteTree(...)` at first glance appears to copy most of the code from `walker.TreeRewriter`. Unnecessarily duplicating code is a massive maintenance burden.

dgkmd committed:
Merge branch 'master' into feature/rechunker

dgkmd committed:
Refactor: rechunker RewriteTree

Make rechunker.RewriteTree() use walker.TreeRewriter
Update rechunker.gatherFileContents() to follow current data.StreamTrees() signature

dgkmd committed:
Misc: Correct style typo

dgkmd committed:
Refactor: Integrate rechunk-copy to copy command

Integrate `rechunk-copy` command to `copy --rechunk` command

dgkmd committed:
Misc(Rechunker): Update comments and names

Minor change of comments and variable names for readability

wplapper commented on changelog/unreleased/issue-5473:
```diff
@@ -0,0 +1,25 @@
+and then backup them again to the new repository. This workaround was 
+inefficient, prone to tamper with the original metadata, and bothersome.
+
+It now supports `rechunk-copy` command, in which the data are rechunked while copying.
```
This file  needs some rework as well, since you have moved the code into the ``restic copy --rechunk`` command.

wplapper commented on cmd/restic/cmd_rechunk_copy.go:
```diff
@@ -0,0 +1,228 @@
+
+// Reference: cmd_copy.go (v0.18.0)
+
+func newRechunkCopyCommand(globalOptions *global.Options) *cobra.Command {
```
According to my linter, newRechunkCopyCommand() is unused. So I assume that there will be more unused functions in here.

dgkmd committed:
Refactor: Helper functions

Change helper function signatures for rechunker, dispatcher and eventTracker
Improve error handling logic for AdvanceCursor

dgkmd committed:
Test: Refactor rechunk-copy integration test

Refactor rechunk-copy integration test from cmd_rechunk_copy_integration_test into cmd_copy_integration_test

dgkmd committed:
Docs: Update rechunk copy documentation

Update changelog and user docs regarding rechunk copy

dgkmd committed:
Refactor: Rechunker

Integrate Dispatcher and eventTracker in rechunker package into Scheduler

dgkmd committed:
Misc: Refine comments for functions

dgkmd committed:
Fix: Integration test for rechunk copy

Fix code redundancy and test ommision in rechunk copy integration test

dgkmd committed:
Fix: Rechunker TreeRewriter behavior

Change rechunker TreeRewriter to allow rewriting unstable json.

dgkmd committed:
Refactor: Move Cursor type definition to worker.go

dgkmd committed:
Misc: Change variable name in rechunker

dgkmd committed:
Test: Refactor rechunker test toward code reuse

Refactor rechunker test code to largely reuse code in other packages, for maintainability

dgkmd committed:
Misc: gofmt

Run gofmt for cmd_copy_integration_test.go

dgkmd committed:
Test: Use t.Run for test in rechunker_test

Use testing.T.Run() scheme for test routines in rechunker_test

dgkmd committed:
Refactor: Rechunker

Cleanup unnecessary pieces of code in rechunker.go

dgkmd committed:
Refactor: Rechunker RewriteTree

Refactor rechunker RewriteTree (including rename to RewriteTrees).

Change cmd_copy.go to align with refactored RewriteTrees.

dgkmd committed:
Test: Rechunker

Compact rechunker test code

wplapper commented:
The ``restic copy --rechunk (--force)`` option is the perfect opportunity to create a fresh copy of a repository, leaving all unused and duplicates blobs behind which are filling up the source repository. I would call this "spring cleaning". Highly recommended on top of all the other features ``--rechunk`` offers.

dgkmd committed:
Refactor: Change rechunker.Index to interface

Change rechunker.Index to interface type, so that the index is expandable to custom types

dgkmd committed:
Refactor: Change function names

Change function names in rechunker to

dgkmd committed:
Refactor: Rename variables and functions

Rename variables and functions for readability
Create WorkerConfig for cleaner argument passing

dgkmd commented:
Hi @wplapper,
I'm glad that you find it useful! Please let me know if something does not run as you expected.

Currently, I'm focusing on refactoring/compacting the code, especially reusing existing code in other packages (following MichaelEischer's comment).
You can find that `Rechunker.RewriteTrees` is reimplemented with `walker.TreeRewriter` and `rechunker_test.go` is simplified with structs from `{archiver, data, walker}` etc.

wplapper commented:
Hi @dgkmd,

I ran a ``restic copy --rechunk --force`` from an existing repository on local disk to an empty repository on local disk with debugging enabled. Here are the basic parameters, as taken for the output ftom the copy command:
```
>/tmp/restic-debug.log; DEBUG_LOG=/tmp/restic-debug.log  ./restic -r /media/user/internal/restic_home_rechunk_copy/ copy --from-repo /media/user/internal/restic_home_master --rechunk --force -vv
debug log file /tmp/restic-debug.log
debug enabled
repository 42fcc4ec opened (version 2, compression level better)
repository ab554fe5 opened (version 2, compression level better)
[0:00] 100.00%  2 / 2 index files loaded
[0:00]          0 index files loaded
Gathering snapshots...
Scanning files to process... 

[Pre-run Summary]
Number of snapshots: 114
Number of distinct files to process: 63691
  - Total size (including duplicate blobs): 16.239 GiB
Number of packs to download: 589

[1:31] 63691/63691 distinct files processed
100.00% 16.239 GiB/16.239 GiB

Rewriting trees...
Rewriting done.

Writing snapshots
snapshot ba4982a6 saved, copied from source snapshot 5fb727a5
... repeat 113 times ...
Additional data stored to the repository: 9.050 GiB
```

The timings (extracted from RESTIC_DEBUG log):
```
2026/04/28 11:18:19 === start ===
2026/04/28 11:18:21 restic/cmd_copy.go:409	main.rechunkCopy	1	Running Plan()
2026/04/28 11:19:52 rechunker/rechunker.go:239  rechunker.(*Rechunker).Rechunk  1   Rechunk done. 
2026/04/28 11:19:52 restic/cmd_copy.go:433  main.rechunkCopy.func2  128176  Running RewriteTree() for tree ID 3f0748bf
2026/04/28 11:21:00 repository/packer_manager.go:66 repository.(*packerManager).Flush   128176  manually flushing pending pack
2026/04/28 11:21:01 ... saving snapshots ...
2026/04/28 11:21:01 === ended ===

time for startup, opening repos, planning 00:02
time to rechunk                           01:31
time to rewrite trees                     01:08
time to save snapshots                    00:01
```

I was surprised to find that the rewriting of the trees takes nearly as long as the rechunking. Would it be possible to improve the speed of the ``rewrite trees`` part? Could one transform all trees once and then use ``data.StreamTrees()`` instead of walking all the snapshots individually one after the other? Just a question and an idea. Alternatively you could think about running these `rechnker.RewriteTree()`` concurrently.

The number of the ``rechunk``part is about 1 million lines, and the ``rewrite`` part is a another 1.1 million lines of debugging output. So there is a lot going on while rewriting the trees. The debug output in total is about 333 MiB in size.

dgkmd commented:
Hi @wplapper,
Sorry for the late response, and thank you for the feedback!

It is indeed unusual that it takes long time rewriting trees. I could not reproduce the issue with my repos.

Current `Rechunker.RewriteTrees()` uses `walker.TreeRewriter`, and it remembers previously processed trees in `idMap` so it should not take so long time even when processing multiple snapshots. 
And as rewriting of the upper directory depends on the rewriting result of subdirectories, performance gain by concurrent running might not be much compared with added complexity to handle the dependency.

There are a few things I would like to ask you for debugging:
(1) Is the local cache enabled for the source repo? Restic supports local cache for tree blobs, and TreeRewriter depends on that cache. So if local cache is disabled, it can take long time like when running `Rechunk()` with `BlobCache` disabled.
(2) Did the issue appear after the code is changed to use `walker.TreeRewriter` (commit 342769db0807f96a0e00d792e0dd0e2747711550)? Could you test with commit 8ab9d44f6d934b8a08e1466e3aac489aafeb284f (which does not use `walker.TreeRewriter`) and let me know the result?
(3) Can you let me know the approximate number of directories (corresponding to the number of tree blobs) in all your snapshots? Mine has no more than 1/10th lines of log for `RewriteTrees()` than for `Rechunk()`, so maybe snapshot characteristic might be different between us.

Thanks again for the feedback!

wplapper commented:
Hi dgkmd,

I messed up your copy of the restic code. Sorry about that. I now have downloaded (git clone) a new copy of your repository, compiled it and everything is working as expected. 

Sorry for the mess I have created for you.

dgkmd commented:
Hi @wplapper,
No worries :)
If you find another issue or have some suggestions, please let me know.

dgkmd commented:
Hi @MichaelEischer,
Thank you for reviewing.

Please let me know if there is something I can assist you with regarding this PR.

MichaelEischer commented:
> Please let me know if there is something I can assist you with regarding this PR.

I still haven't got around to make the changes I mentioned a while back, the review for this PR will unfortunately have to wait until then.

dgkmd committed:
Refactor: Feed workgroup to submodules

Feed workgroup argument when creating blob cache and scheduler so that internal errors of those submodules' background workers can be captured by the outer workflow
Add BlobCacheConfig type to wrap blob cache parameters

dgkmd commented:
Hi @MichaelEischer ,

I'm tracking your PR #21943.
Please let me know if your feature is ready or there is something I can do for you.

MichaelEischer commented:
@dgkmd I'll let you know when I'm far enough to get to this PR, sorry for taking so long
